### DISCUSSION TITLE
New instance can't search for or follow remote accounts

### DISCUSSION THREAD

**Discussion by abbottmg** [mastodon/mastodon#20257] [Q&A] [answered]
Hi all,

I just got a server setup last night and while accounts can see and follow other local accounts, interacting with other instances fails. Searching for `@KNOWN_USER@mastodon.social` returns an error toast "503 remote data could not be fetched". Another user initiated a follow and the job keeps retrying and failing on sidekiq:

````
2022-11-09T16:06:51.404Z pid=7 tid=2flj class=ActivityPub::DeliveryWorker jid=d631ebb5839a828f727cc0ec INFO: start
2022-11-09T16:06:51.418Z pid=7 tid=2flj class=ActivityPub::DeliveryWorker jid=d631ebb5839a828f727cc0ec elapsed=0.014 INFO: fail
2022-11-09T16:06:51.419Z pid=7 tid=2flj WARN: {"context":"Job raised exception","job":{"retry":16,"queue":"push","dead":false,"args":["{\"@context\
":\"https://www.w3.org/ns/activitystreams\",\"id\":\"https://oliphaunt.social/21b04235-6f7b-4a64-aca5-7d9f60c74d7e\",\"type\":\"Follow\",\"actor\":
\"https://oliphaunt.social/users/Mithradatesvi\",\"object\":\"\"}",109313995065247520,""],"class":"ActivityPub::DeliveryWorker","jid":"d631ebb5839a
828f727cc0ec","created_at":1668008835.768326,"enqueued_at":1668010011.4041595,"error_message":"ArgumentError","error_class":"ArgumentError","failed
_at":1668008835.8052967,"retry_count":5,"retried_at":1668009370.043162},"jobstr":"{\"retry\":16,\"queue\":\"push\",\"dead\":false,\"args\":[\"{\\\"
@context\\\":\\\"https://www.w3.org/ns/activitystreams\\\",\\\"id\\\":\\\"https://oliphaunt.social/21b04235-6f7b-4a64-aca5-7d9f60c74d7e\\\",\\\"typ
e\\\":\\\"Follow\\\",\\\"actor\\\":\\\"https://oliphaunt.social/users/Mithradatesvi\\\",\\\"object\\\":\\\"\\\"}\",109313995065247520,\"\"],\"class
\":\"ActivityPub::DeliveryWorker\",\"jid\":\"d631ebb5839a828f727cc0ec\",\"created_at\":1668008835.768326,\"enqueued_at\":1668010011.4041595,\"error
_message\":\"ArgumentError\",\"error_class\":\"ArgumentError\",\"failed_at\":1668008835.8052967,\"retry_count\":5,\"retried_at\":1668009370.043162}
"}
2022-11-09T16:06:51.419Z pid=7 tid=2flj WARN: ArgumentError: ArgumentError
2022-11-09T16:06:51.419Z pid=7 tid=2flj WARN: /opt/mastodon/app/lib/request.rb:28:in `initialize'
/opt/mastodon/app/workers/activitypub/delivery_worker.rb:39:in `new'
/opt/mastodon/app/workers/activitypub/delivery_worker.rb:39:in `build_request'
````

I'm not familiar with the system yet but I don't actually see the follow target listed in that big warning object, so this may be unrelated.

I'm using the docker-compose setup; `docker logs mastodon_web_1` doesn't return any detail beyond the 503; I'll be looking into upgrading the log level here shortly.

````
[ba8d240c-54e5-457e-adbc-6c90d85b23e8] method=GET path=/api/v2/search format=html controller=Api::V2::SearchController action=index status=503 duration=10444.23 view=0.80 db=12.52
[f6e0f5f1-6fa5-4e69-88b5-78a059de28db] method=HEAD path=/health format=*/* controller=HealthController action=show status=200 duration=0.56 view=0.29 db=0.00
[ActionDispatch::HostAuthorization::DefaultResponseApp] Blocked host: 45.79.10.231
[ActionDispatch::HostAuthorization::DefaultResponseApp] Blocked host: 45.79.10.231
````

The blocked host lines do have my server's IP so I included them in case we'd expect another server to try to phone back by IP or something. I think they're just a coincidence though, as other examples of search failing don't have any similar lines.

I tried to keep the server config as on-rails as possible, with the biggest change being the addition of nginx as a containerized service in `docker-compose.yml` rather than a service on the host machine. I'm happy to provide any configs etc. if they'll help.

**Comment by abbottmg** [comment]
Oh, and I also can't hit the /sidekiq endpoint, which the [tutorial I followed](https://www.linode.com/docs/guides/install-mastodon-on-ubuntu-2004/#download-mastodon) says I should see. The request makes it to the app server, which returns a 404.

When I didn't see anything about sidekiq in `dist/nginx.conf` I figured the app server was forwarding that endpoint. I'm happy to have nginx handle that location but is there a best practice for only letting admin users see it or is that even something to worry about?

**Comment by abbottmg** [comment]
Aaaand I'm going to include another issue in here, because this could either be a reachability bug or a misconfiguration of object storage on my part (or neither).

I uploaded an avatar for my account and failed twice. It looks like it's receiving the multipart and converting it successfully, but then saving it to the /accounts/avatars path fails:

````
[d9cf223d-a427-4a00-8912-6f2da63241a2] [paperclip] Trying to link /tmp/RackMultipart20221109-19-q6rugu.jpeg to /tmp/407c952e1329c7ce26587cf28e6efce720221109-19-pf5m6q.jpeg
[d9cf223d-a427-4a00-8912-6f2da63241a2] [paperclip] Trying to link /tmp/407c952e1329c7ce26587cf28e6efce720221109-19-pf5m6q.jpeg to /tmp/7c51d7ad00c47e1c02bd03409fe6436020221109-19-p2ea8n.jpeg
[d9cf223d-a427-4a00-8912-6f2da63241a2] Command :: file -b --mime '/tmp/7c51d7ad00c47e1c02bd03409fe6436020221109-19-p2ea8n.jpeg'
[d9cf223d-a427-4a00-8912-6f2da63241a2] Command :: identify -format %m '/tmp/407c952e1329c7ce26587cf28e6efce720221109-19-pf5m6q.jpeg[0]'
[d9cf223d-a427-4a00-8912-6f2da63241a2] Command :: convert '/tmp/407c952e1329c7ce26587cf28e6efce720221109-19-pf5m6q.jpeg[0]' -auto-orient -resize "400x" -crop "400x400+0+66" +repage -strip '/tmp/79517c0008c85dc92b67da959ff208d920221109-19-7kai36'
[d9cf223d-a427-4a00-8912-6f2da63241a2] [paperclip] Trying to link /tmp/79517c0008c85dc92b67da959ff208d920221109-19-7kai36 to /tmp/92bdb361c5bc10b218f51ecf6cf7bb3d20221109-19-cs91hl
[d9cf223d-a427-4a00-8912-6f2da63241a2] [paperclip] saving accounts/avatars/109/314/550/281/205/695/original/03cd7615c5633c56.jpeg
[d9cf223d-a427-4a00-8912-6f2da63241a2] Storage server error: execution expired
[d9cf223d-a427-4a00-8912-6f2da63241a2] method=PUT path=/settings/profile format=html controller=Settings::ProfilesController action=update status=503 duration=10136.53 view=1.84 db=3.24
````

**Comment by iamgitcat** [comment]
Hi. You should try "KNOWN_USER@mastodon.social". That's the username, without @.

**Reply by abbottmg** [reply]
Just double checked and that also fails unfortunately. As does copying the link from another user's follow dialog.

I was able to find my test user once and follow (immediately after I rebooted my server), but other users both on mastodon.social and mstdn.social are still failing.

**Comment by abbottmg** [comment, marked as answer]
Believe it or not, the two issues _were_ linked after all! When searching for a remote user (including when directly attempting to follow them, apparently), mastodon fetches their avatar and banner into AR alongside the "actually required" account information. When AR goes to save, paperclip tries to push that media to the cache in S3, and if that's not properly configured, the whole save fails and a 503 is returned as if the user was corrupted.

Below are the settings that eventually worked for me (at this point I'm not sure what the original settings were). This might be useful for anyone else who's using this project as an excuse to learn multiple pieces of the ops puzzle along the way:

````
S3_ENABLED=true
S3_PROTOCOL=https                # or http until you learn how to generate a cert and hand it to your S3 provider
S3_BUCKET=files.oliphaunt.social # examples recommend matching this to my host but I'm not sure it mattered for me
S3_REGION=default                # linode often recommends an empty string but conf/initializers/paperclip.rb will fill that in with an AWS region
S3_HOSTNAME=us-southeast-1.linodeobjects.com
S3_ENDPOINT=https://us-southeast-1.linodeobjects.com # KEY: contrary to other examples I COULD NOT include the bucket name here
AWS_ACCESS_KEY_ID=(secret)
AWS_SECRET_ACCESS_KEY=(secret)
S3_ALIAS_HOST=files.oliphaunt.social # the subdomain that will need coverage by the aforementioned cert
S3_ALIAS_PREFIXES=1                  # intended to strip the bucket name from the path
S3_OVERRIDE_PATH_STYLE=true          # put bucket name as subdomain of S3_HOSTNAME rather than in the path
````

I added the `S3_ALIAS_PREFIXES` key, which is read if `S3_ALIAS_HOST` exists and is optional with a default of 0. I'm going to find a low-traffic moment to test removing it because I'm not sure it's actually necessary with `S3_OVERRIDE_PATH_STYLE` set. If it is, I'll wrap the new setting up in a PR.

While iterating on these settings I had found cases where the files were uploaded with the bucket name in both the subdomain position **and** as a path component, basically putting everything one "folder" level deeper; the fetch process was then looking things up **without** the bucket in the path and failing to find.
